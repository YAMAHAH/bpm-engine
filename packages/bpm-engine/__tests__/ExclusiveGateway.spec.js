import fs from 'fs';

import BPMEngine from 'bpm-engine';

describe('ExclusiveGateway', () => {
  it('Does not continue if no DefaultSequenceFlow was found', async () => {
    const bpm = new BPMEngine();

    const token = await bpm.createProcessInstance({
      workflowDefinition: fs.readFileSync(
        `${__dirname}/diagrams/ExclusiveGatewayWithoutDefaultSequenceFlow.bpmn`,
        'utf-8',
      ),
    });

    await token.exec();
    expect(token.next).toMatchSnapshot();
  });

  it('Continues to the evaluated conditions paths', async () => {
    const passedTasks = [];

    class TaskPlugin extends BPMEngine.Plugins.Element {
      // eslint-disable-next-line
      onActive(definition) {
        if (definition.id === 'Task_02ex5w4') {
          passedTasks.push(definition.id);
        }
        if (definition.id === 'Task_1tasl7x') {
          passedTasks.push(definition.id);
        }
      }
    }

    const taskPlugin = new TaskPlugin();

    const bpm = new BPMEngine({
      plugins: [taskPlugin],
    });

    const tiTop = await bpm.createProcessInstance({
      workflowDefinition: fs.readFileSync(`${__dirname}/diagrams/ExclusiveGateway.bpmn`, 'utf-8'),
      payload: {
        top: true,
      },
    });

    const tiBottom = await bpm.createProcessInstance({
      workflowDefinition: fs.readFileSync(`${__dirname}/diagrams/ExclusiveGateway.bpmn`, 'utf-8'),
      payload: {
        bottom: true,
      },
    });

    const tiDefault = await bpm.createProcessInstance({
      workflowDefinition: fs.readFileSync(`${__dirname}/diagrams/ExclusiveGateway.bpmn`, 'utf-8'),
      payload: {},
    });

    await tiTop.exec();
    await tiBottom.exec();
    await tiDefault.exec();

    expect(tiTop.next).toMatchSnapshot();
    expect(tiBottom.next).toMatchSnapshot();
    expect(tiDefault.next).toMatchSnapshot();
    expect(passedTasks).toMatchSnapshot();
  });
});
