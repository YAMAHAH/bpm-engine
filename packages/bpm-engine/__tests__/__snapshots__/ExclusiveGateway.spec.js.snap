// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExclusiveGateway Continues to the evaluated conditions paths 1`] = `
Base {
  "$type": "bpmn:EndEvent",
  "id": "EndEvent_1915ikh",
  "name": "End",
}
`;

exports[`ExclusiveGateway Continues to the evaluated conditions paths 2`] = `
Base {
  "$type": "bpmn:EndEvent",
  "id": "EndEvent_1915ikh",
  "name": "End",
}
`;

exports[`ExclusiveGateway Continues to the evaluated conditions paths 3`] = `
Base {
  "$type": "bpmn:EndEvent",
  "id": "EndEvent_1915ikh",
  "name": "End",
}
`;

exports[`ExclusiveGateway Continues to the evaluated conditions paths 4`] = `
Array [
  "Task_02ex5w4",
  "Task_1tasl7x",
  "Task_1tasl7x",
]
`;

exports[`ExclusiveGateway Does not continue if no DefaultSequenceFlow was found 1`] = `
Base {
  "$type": "bpmn:ExclusiveGateway",
  "id": "ExclusiveGateway_1kcsj1f",
}
`;
